zephyr/esp32/src/wifi/esp_wifi_adapter.c

```diff
@@ -37,6 +37,36 @@ LOG_MODULE_REGISTER(esp32_wifi_adapter, CONFIG_WIFI_LOG_LEVEL);
 
 ESP_EVENT_DEFINE_BASE(WIFI_EVENT);
 
+#ifdef CONFIG_ESP32_WIFI_HEAP_MEM_POOL
+
+#define TAG "wifi"
+
+/* ESP heap extension */
+extern unsigned z_mapped_end;
+extern unsigned _heap_sentry;
+
+static void *esp_wifi_heap_init_mem = &z_mapped_end;
+#define ESP_WIFI_HEAP_SIZE  ((uintptr_t)&_heap_sentry - (uintptr_t)&z_mapped_end)
+
+static struct k_heap esp_wifi_heap;
+
+static int esp_wifi_heap_init(void)
+{
+	ESP_EARLY_LOGI(TAG, "ESP Wi-Fi heap init at 0x%x size %d kB.\n",
+			esp_wifi_heap_init_mem, ESP_WIFI_HEAP_SIZE/1024);
+	k_heap_init(&esp_wifi_heap, esp_wifi_heap_init_mem, ESP_WIFI_HEAP_SIZE);
```

Comment:
What if the extension does not have enough memory? Can we check and use sys_heap instead?

Reply:
we don't know the `z_mapped_end` until the link phase is done.

Reply:
> Can we check and use sys_heap instead?

Yes, we could add some kind of runtime mempool switching feature to the adapter to do this. Let's add this feature in another PR.